src/neptune_scale/util/source_tracking.py

```diff
@@ -0,0 +1,184 @@
+        remotes=remotes,
+        entry_point_path=entry_point_path,
+        head_diff_content=head_diff_content,
+        upstream_diff_commit_id=upstream_sha,
+        upstream_diff_content=upstream_diff_content,
+    )
+
+
+def _get_git_repo(path: Optional[pathlib.Path]) -> Optional["git.Repo"]:
+    if path is not None:
+        path = path.resolve()
+
+    try:
+        import git
+
+        try:
+            return git.Repo(path, search_parent_directories=True)
+        except (git.exc.NoSuchPathError, git.exc.InvalidGitRepositoryError):
+            return None
+    except ImportError:
+        logger.warn("GitPython could not be initialized")
+        return None
+
+
+def _get_active_branch(repo: "git.Repo") -> str:
+    active_branch = ""
+    try:
+        active_branch = repo.active_branch.name
+    except TypeError as e:
+        if str(e.args[0]).startswith("HEAD is a detached symbolic reference as it points to"):
+            active_branch = "Detached HEAD"
+    return active_branch
+
+
+def _read_diff(repo: "git.Repo", commit_ref: str) -> Optional[bytes]:
+    try:
+        from git.exc import GitCommandError
+
+        try:
+            diff = repo.git.diff(commit_ref, index=False)
+
+            if not isinstance(diff, str):
+                return None
+
+            # add a newline at the end (required to be a valid `patch` file)
+            if diff and diff[-1] != "\n":
+                diff += "\n"
+
+            return diff.encode("utf-8")
+        except GitCommandError:
+            return None
+    except ImportError:
+        return None
+
+
+def _get_relevant_upstream_commit(repo: "git.Repo") -> Optional["git.Commit"]:
+    try:
+        tracking_branch = repo.active_branch.tracking_branch()
+    except (TypeError, ValueError):
+        return None
+
+    if tracking_branch:
+        return tracking_branch.commit
+
+    return _search_for_most_recent_ancestor(repo)
+
+
+def _search_for_most_recent_ancestor(repo: "git.Repo") -> Optional["git.Commit"]:
+    most_recent_ancestor: Optional[git.Commit] = None
+
+    try:
+        from git.exc import GitCommandError
+
+        try:
+            for branch in repo.heads:
+                tracking_branch = branch.tracking_branch()
+                if tracking_branch:
+                    for ancestor in repo.merge_base(repo.head, tracking_branch.commit):
+                        if not most_recent_ancestor or repo.is_ancestor(most_recent_ancestor, ancestor):
+                            most_recent_ancestor = ancestor
+        except GitCommandError:
+            pass
+    except ImportError:
+        return None
+
+    return most_recent_ancestor
+
+
+def _get_entrypoint_path() -> Optional[pathlib.Path]:
+    if _is_ipython():
+        return None
+
+    entrypoint_path = get_path_executed_script()
+    if not isinstance(entrypoint_path, pathlib.Path):
+        return None
+    if entrypoint_path == pathlib.Path() or not entrypoint_path.is_file():
```

Comment:
Not sure, but I suspect the first condition is redundant. `pathlib.Path().is_file()` is probably always false.